File: content/problems/2908-minimum-sum-of-mountain-triplets-i/solutions/running-left-and-suffix-minimum/solution.cpp
class Solution {
    public:int minimumSum(vector<int>& nums){
        int n=nums.size();
        vector<int> rightMinimum(n);
        rightMinimum[n-1]=nums[n-1];
        for(int i=n-2;i>=0;--i)rightMinimum[i]=min(nums[i],rightMinimum[i+1]);
        int leftMinimum=nums[0],answer=INT_MAX;
        for(int j=1;j<n-1;++j){
            if(leftMinimum<nums[j]&&rightMinimum[j+1]<nums[j])answer=min(answer,leftMinimum+nums[j]+rightMinimum[j+1]);
            leftMinimum=min(leftMinimum,nums[j]);
        }
        return answer==INT_MAX?-1:answer;
    }
};

File: content/problems/2908-minimum-sum-of-mountain-triplets-i/solutions/running-left-and-suffix-minimum/Solution.kt
class Solution {
    fun minimumSum(nums:IntArray):Int{
        val n=nums.size;
        val rightMinimum=IntArray(n);
        rightMinimum[n-1]=nums[n-1];
        for(i in n-2 downTo 0)rightMinimum[i]=minOf(nums[i],rightMinimum[i+1]);
        var leftMinimum=nums[0];
        var answer=Int.MAX_VALUE;
        for(j in 1 until n-1){
            if(leftMinimum<nums[j]&&rightMinimum[j+1]<nums[j])answer=minOf(answer,leftMinimum+nums[j]+rightMinimum[j+1]);
            leftMinimum=minOf(leftMinimum,nums[j])
        };
        return if(answer==Int.MAX_VALUE)-1 else answer
    }
}
